### DISCUSSION TITLE
clang can not find libz3.so.4.11

### DISCUSSION THREAD

**Discussion by cang-mang** [Homebrew/discussions#3934] [Linux] [answered]
Today I run `brew upgrade` and the llvm is upgraded. After everything is finished, I run `clang --version`, but it reports `clang: error while loading shared libraries: libz3.so.4.11: cannot open shared object file: No such file or directory`.

**Comment by SMillerDev** [comment]
What does `brew doctor` and `brew info llvm` say?

**Reply by cang-mang** [reply]
```
$brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  linux-headers@5.15

Warning: Formulae which link to GCC through a versioned path were found. These formulae
are prone to breaking when GCC is updated. You should `brew reinstall` these formulae:
  gcc@11
```

**Reply by cang-mang** [reply]
```
$brew info llvm
==> llvm: stable 15.0.4 (bottled), HEAD
Next-gen compiler infrastructure
https://llvm.org/
/home/linuxbrew/.linuxbrew/Cellar/llvm/15.0.4 (6,512 files, 1.9GB) *
  Poured from bottle on 2022-11-17 at 17:39:42
From: https://mirrors.tuna.tsinghua.edu.cn/git/homebrew/homebrew-core.git/Formula/llvm.rb
License: Apache-2.0 with LLVM-exception
==> Dependencies
Build: cmake ✘, swig ✘, pkg-config ✔
Required: python@3.11 ✔, six ✔, z3 ✔, zstd ✔, libedit ✔, libffi ✔, ncurses ✔, zlib ✔, binutils ✔, elfutils ✔, gcc ✔, glibc ✔
==> Options
--HEAD
	Install HEAD version
==> Caveats
Emacs Lisp files have been installed to:
  /home/linuxbrew/.linuxbrew/share/emacs/site-lisp/llvm
==> Analytics
install: 59,111 (30 days), 156,488 (90 days), 483,080 (365 days)
install-on-request: 48,043 (30 days), 120,706 (90 days), 336,260 (365 days)
build-error: 473 (30 days)
```

**Comment by carlocab** [comment, marked as answer]
Do `brew reinstall z3` then try again.

**Reply by cang-mang** [reply]
It does not work. Because `libz3.so.4.11` is not in `/home/linuxbrew/.linuxbrew/Cellar/z3/4.11.2/lib`.

```
$ls -F /home/linuxbrew/.linuxbrew/Cellar/z3/4.11.2/lib
libz3.a  libz3.so*  python3.10/
```

**Reply by carlocab** [reply]
It should be there upon reinstall, though. Try
```
brew update && brew reinstall z3
```
If that doesn't help, what's the output of `brew config`?

**Reply by cang-mang** [reply]
Today I try `brew reinstall z3` again. It works. Thank you!

**Comment by Blaisorblade** [comment]
TLDR: Same issue, `reinstall` helped, but is it a Homebrew bug that it didn't upgrade the bottle in 2 months?

- I needed the same solution; I had the same problem and brew tried to rebuild llvm from source.

- Worringly, Homebrew didn't notice it needed to upgrade the bottle for 2 months? Before the reinstall, `brew info z3` showed the bottle was from September, before @carlocab 's https://github.com/Homebrew/homebrew-core/pull/111079 and before the bottle change in https://github.com/Homebrew/homebrew-core/commit/831455c776566517121859c271532d102d675b83.

```
$ brew info z3
==> z3: stable 4.11.2 (bottled), HEAD
High-performance theorem prover
https://github.com/Z3Prover/z3
/opt/homebrew/Cellar/z3/4.11.2 (136 files, 97.2MB) *
  Poured from bottle on 2022-09-19 at 13:10:04
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/z3.rb
License: MIT
==> Dependencies
Build: cmake ✔, python@3.10 ✔
==> Options
--HEAD
	Install HEAD version
==> Analytics
install: 31,712 (30 days), 94,132 (90 days), 106,996 (365 days)
install-on-request: 1,787 (30 days), 4,162 (90 days), 12,072 (365 days)
build-error: 3,900 (30 days)
```

```
$ brew info z3
==> z3: stable 4.11.2 (bottled), HEAD
High-performance theorem prover
https://github.com/Z3Prover/z3
/opt/homebrew/Cellar/z3/4.11.2 (136 files, 97.2MB) *
  Poured from bottle on 2022-09-19 at 13:10:04
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/z3.rb
License: MIT
==> Dependencies
Build: cmake ✔, python@3.10 ✔
==> Options
--HEAD
	Install HEAD version
==> Analytics
install: 31,712 (30 days), 94,132 (90 days), 106,996 (365 days)
install-on-request: 1,787 (30 days), 4,162 (90 days), 12,072 (365 days)
build-error: 3,900 (30 days)
```

**Comment by goberlan** [comment]
This issue still happens... I tried the reinstall multiple times, does not fix it. I tried `brew update && brew reinstall z3` does not work

**Reply by p-linnane** [reply]
> This issue still happens... I tried the reinstall multiple times, does not fix it. I tried `brew update && brew reinstall z3` does not work

Can't reproduce this.

```
╰─ /opt/homebrew/Cellar/llvm/20.1.7/bin/clang --version
Homebrew clang version 20.1.7
Target: arm64-apple-darwin25.0.0
Thread model: posix
InstalledDir: /opt/homebrew/Cellar/llvm/20.1.7/bin
Configuration file: /opt/homebrew/etc/clang/arm64-apple-darwin25.cfg
```

```
╰─ brew linkage llvm
System libraries:
  /System/Library/Frameworks/CoreFoundation.framework/Versions/A/CoreFoundation
  /System/Library/Frameworks/CoreServices.framework/Versions/A/CoreServices
  /System/Library/Frameworks/Foundation.framework/Versions/C/Foundation
  /System/Library/Frameworks/Security.framework/Versions/A/Security
  /usr/lib/libSystem.B.dylib
  /usr/lib/libc++.1.dylib
  /usr/lib/libc++abi.dylib
  /usr/lib/libcompression.dylib
  /usr/lib/libedit.3.dylib
  /usr/lib/libffi.dylib
  /usr/lib/libform.5.4.dylib
  /usr/lib/libncurses.5.4.dylib
  /usr/lib/libobjc.A.dylib
  /usr/lib/libpanel.5.4.dylib
  /usr/lib/libxml2.2.dylib
  /usr/lib/libz.1.dylib
Homebrew libraries:
  /opt/homebrew/Cellar/llvm/20.1.7/lib/c++/libc++abi.1.dylib (llvm)
  /opt/homebrew/Cellar/llvm/20.1.7/lib/libLLVM.dylib (llvm)
  /opt/homebrew/Cellar/llvm/20.1.7/lib/libMLIR.dylib (llvm)
  /opt/homebrew/Cellar/llvm/20.1.7/lib/libclang-cpp.dylib (llvm)
  /opt/homebrew/Cellar/llvm/20.1.7/lib/libclang.dylib (llvm)
  /opt/homebrew/Cellar/llvm/20.1.7/lib/liblldb.20.1.7.dylib (llvm)
  /opt/homebrew/Cellar/llvm/20.1.7/lib/libmlir_float16_utils.dylib (llvm)
  /opt/homebrew/Cellar/llvm/20.1.7/lib/unwind/libunwind.1.dylib (llvm)
  /opt/homebrew/opt/python@3.13/Frameworks/Python.framework/Versions/3.13/Python (python@3.13)
  /opt/homebrew/opt/xz/lib/liblzma.5.dylib (xz)
  /opt/homebrew/opt/z3/lib/libz3.4.15.dylib (z3)
  /opt/homebrew/opt/zstd/lib/libzstd.1.dylib (zstd)
```
